File: sdk-go/sign.go
package banhbaoring

import (
	"context"
	"encoding/base64"
	"fmt"

	"github.com/google/uuid"
)

// SignService handles signing operations.
type SignService struct {
	client *Client
}

// SignRequest is the request for signing data.
type SignRequest struct {
	// KeyID is the ID of the key to sign with.
	KeyID uuid.UUID `json:"key_id"`
	// Data is the data to sign (will be base64-encoded for transmission).
	Data []byte `json:"-"`
	// Prehashed indicates if the data is already hashed.
	Prehashed bool `json:"prehashed,omitempty"`
}

// SignResponse is the response from a sign operation.
type SignResponse struct {
	// KeyID is the ID of the key that signed.
	KeyID uuid.UUID `json:"key_id"`
	// Signature is the raw signature bytes.
	Signature []byte `json:"-"`
	// PublicKey is the hex-encoded public key.
	PublicKey string `json:"public_key"`
	// KeyVersion is the version of the key that signed.
	KeyVersion int `json:"key_version"`
}

// BatchSignRequest signs multiple messages in parallel.
type BatchSignRequest struct {
	// Requests is the list of sign requests.
	Requests []SignRequest `json:"requests"`
}

// BatchSignResult is a single result from a batch sign operation.
type BatchSignResult struct {
	// KeyID is the ID of the key that signed.
	KeyID uuid.UUID `json:"key_id"`
	// Signature is the raw signature bytes, or nil if there was an error.
	Signature []byte `json:"-"`
	// PublicKey is the hex-encoded public key.
	PublicKey string `json:"public_key,omitempty"`
	// KeyVersion is the version of the key that signed.
	KeyVersion int `json:"key_version,omitempty"`
	// Error is the error message if signing failed for this request.
	Error string `json:"error,omitempty"`
}

// Sign signs data with a key.
//
// Example:
//
//	result, err := client.Sign.Sign(ctx, keyID, []byte("message to sign"), false)
//	signature := result.Signature
func (s *SignService) Sign(ctx context.Context, keyID uuid.UUID, data []byte, prehashed bool) (*SignResponse, error) {
	req := map[string]interface{}{
		"data":      base64.StdEncoding.EncodeToString(data),
		"prehashed": prehashed,
	}

	var resp struct {
		Signature  string `json:"signature"`
		PublicKey  string `json:"public_key"`
		KeyVersion int    `json:"key_version"`
	}

	if err := s.client.post(ctx, fmt.Sprintf("/v1/keys/%s/sign", keyID), req, &resp); err != nil {
		return nil, err
	}

	sig, err := base64.StdEncoding.DecodeString(resp.Signature)
	if err != nil {
		return nil, fmt.Errorf("invalid signature encoding: %w", err)
	}

	return &SignResponse{
		KeyID:      keyID,
		Signature:  sig,
		PublicKey:  resp.PublicKey,
		KeyVersion: resp.KeyVersion,
	}, nil
}

// SignBatch signs multiple messages in parallel.
// This is critical for Celestia's parallel blob submission pattern.
//
// Example:
//
//	results, err := client.Sign.SignBatch(ctx, banhbaoring.BatchSignRequest{
//	    Requests: []banhbaoring.SignRequest{
//	        {KeyID: worker1, Data: tx1},
//	        {KeyID: worker2, Data: tx2},
//	        {KeyID: worker3, Data: tx3},
//	        {KeyID: worker4, Data: tx4},
//	    },
//	})
//	// All 4 sign in parallel - completes in ~200ms, not 800ms!
func (s *SignService) SignBatch(ctx context.Context, req BatchSignRequest) ([]*BatchSignResult, error) {
	// Convert to API format
	requests := make([]map[string]interface{}, len(req.Requests))
	for i, r := range req.Requests {
		requests[i] = map[string]interface{}{
			"key_id":    r.KeyID.String(),
			"data":      base64.StdEncoding.EncodeToString(r.Data),
			"prehashed": r.Prehashed,
		}
	}

	apiReq := map[string]interface{}{
		"requests": requests,
	}

	var resp struct {
		Signatures []struct {
			KeyID      string `json:"key_id"`
			Signature  string `json:"signature,omitempty"`
			PublicKey  string `json:"public_key,omitempty"`
			KeyVersion int    `json:"key_version,omitempty"`
			Error      string `json:"error,omitempty"`
		} `json:"signatures"`
		Count int `json:"count"`
	}

	if err := s.client.post(ctx, "/v1/sign/batch", apiReq, &resp); err != nil {
		return nil, err
	}

	results := make([]*BatchSignResult, len(resp.Signatures))
	for i, sig := range resp.Signatures {
		keyID, _ := uuid.Parse(sig.KeyID)
		result := &BatchSignResult{
			KeyID:      keyID,
			PublicKey:  sig.PublicKey,
			KeyVersion: sig.KeyVersion,
			Error:      sig.Error,
		}

		if sig.Error == "" && sig.Signature != "" {
			sigBytes, err := base64.StdEncoding.DecodeString(sig.Signature)
			if err == nil {
				result.Signature = sigBytes
			}
		}

		results[i] = result
	}

	return results, nil
}

// SignWithOptions provides additional signing options.
type SignWithOptions struct {
	// Prehashed indicates if the data is already hashed.
	Prehashed bool
}

File: sdk-rust/src/sign.rs
//! Signing operations.
//!
//! This module provides the SignClient for signing data with keys stored
//! in BanhBaoRing. Supports both single and batch signing operations.

use crate::client::Client;
use crate::error::{BanhBaoRingError, Result};
use crate::types::{BatchSignRequest, SignResponse};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Client for signing operations.
///
/// Access via `client.sign()`.
pub struct SignClient {
    client: Client,
}

impl SignClient {
    pub(crate) fn new(client: Client) -> Self {
        Self { client }
    }

    /// Sign data with a key.
    ///
    /// # Arguments
    ///
    /// * `key_id` - The key ID to sign with
    /// * `data` - The data to sign
    /// * `prehashed` - If true, data is already hashed (SHA-256)
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let key_id = Uuid::parse_str("...")?;
    ///     
    ///     // Sign raw data (will be hashed by the server)
    ///     let data = b"hello world";
    ///     let result = client.sign().sign(&key_id, data, false).await?;
    ///     
    ///     println!("Signature: {} bytes", result.signature.len());
    ///     println!("Public key: {}", result.public_key);
    ///     Ok(())
    /// }
    /// ```
    ///
    /// # Example with prehashed data
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    /// use sha2::{Sha256, Digest};
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let key_id = Uuid::parse_str("...")?;
    ///     
    ///     // Hash data locally (requires sha2 crate)
    ///     let data = b"hello world";
    ///     let hash = Sha256::digest(data);
    ///     
    ///     // Sign the hash
    ///     let result = client.sign().sign(&key_id, &hash, true).await?;
    ///     Ok(())
    /// }
    /// ```
    pub async fn sign(&self, key_id: &Uuid, data: &[u8], prehashed: bool) -> Result<SignResponse> {
        #[derive(Serialize)]
        struct Request {
            data: String,
            prehashed: bool,
        }

        #[derive(Deserialize)]
        struct Response {
            signature: String,
            public_key: String,
        }

        let request = Request {
            data: BASE64.encode(data),
            prehashed,
        };

        let response: Response = self
            .client
            .post(&format!("/v1/keys/{}/sign", key_id), &request)
            .await?;

        let signature = BASE64
            .decode(&response.signature)
            .map_err(|e| BanhBaoRingError::Decode(e.to_string()))?;

        Ok(SignResponse {
            key_id: *key_id,
            signature,
            public_key: response.public_key,
        })
    }

    /// Sign multiple messages in parallel.
    ///
    /// This is critical for Celestia's parallel blob submission pattern.
    /// All signing operations are performed in a single API call, reducing
    /// latency significantly.
    ///
    /// # Arguments
    ///
    /// * `request` - Batch sign request containing multiple sign items
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::{Client, types::{BatchSignRequest, BatchSignItem}};
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     
    ///     let worker1 = Uuid::parse_str("...")?;
    ///     let worker2 = Uuid::parse_str("...")?;
    ///     let worker3 = Uuid::parse_str("...")?;
    ///     let worker4 = Uuid::parse_str("...")?;
    ///     
    ///     let results = client.sign().sign_batch(BatchSignRequest {
    ///         requests: vec![
    ///             BatchSignItem { key_id: worker1, data: b"tx1".to_vec(), prehashed: false },
    ///             BatchSignItem { key_id: worker2, data: b"tx2".to_vec(), prehashed: false },
    ///             BatchSignItem { key_id: worker3, data: b"tx3".to_vec(), prehashed: false },
    ///             BatchSignItem { key_id: worker4, data: b"tx4".to_vec(), prehashed: false },
    ///         ],
    ///     }).await?;
    ///     
    ///     // All 4 sign in parallel - completes in ~200ms, not 800ms!
    ///     println!("Signed {} transactions", results.len());
    ///     Ok(())
    /// }
    /// ```
    pub async fn sign_batch(&self, request: BatchSignRequest) -> Result<Vec<SignResponse>> {
        #[derive(Serialize)]
        struct ApiRequest {
            requests: Vec<ApiRequestItem>,
        }

        #[derive(Serialize)]
        struct ApiRequestItem {
            key_id: Uuid,
            data: String,
            prehashed: bool,
        }

        #[derive(Deserialize)]
        struct ApiResponse {
            signatures: Vec<ApiSignature>,
        }

        #[derive(Deserialize)]
        struct ApiSignature {
            key_id: Uuid,
            signature: String,
            public_key: String,
            error: Option<String>,
        }

        let api_request = ApiRequest {
            requests: request
                .requests
                .iter()
                .map(|r| ApiRequestItem {
                    key_id: r.key_id,
                    data: BASE64.encode(&r.data),
                    prehashed: r.prehashed,
                })
                .collect(),
        };

        let response: ApiResponse = self.client.post("/v1/sign/batch", &api_request).await?;

        let mut results = Vec::new();
        let mut errors = 0;

        for sig in response.signatures {
            if sig.error.is_some() {
                errors += 1;
                continue;
            }

            let signature = BASE64
                .decode(&sig.signature)
                .map_err(|e| BanhBaoRingError::Decode(e.to_string()))?;

            results.push(SignResponse {
                key_id: sig.key_id,
                signature,
                public_key: sig.public_key,
            });
        }

        if errors > 0 && results.is_empty() {
            return Err(BanhBaoRingError::BatchPartialFailure {
                failed: errors,
                total: request.requests.len(),
            });
        }

        Ok(results)
    }

    /// Verify a signature against the public key of a key.
    ///
    /// # Arguments
    ///
    /// * `key_id` - The key ID to verify with
    /// * `data` - The original data that was signed
    /// * `signature` - The signature to verify
    /// * `prehashed` - If true, data was prehashed before signing
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use banhbaoring::Client;
    /// use uuid::Uuid;
    ///
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     let client = Client::new("bbr_live_xxxxx");
    ///     let key_id = Uuid::parse_str("...")?;
    ///     
    ///     let data = b"hello world";
    ///     let result = client.sign().sign(&key_id, data, false).await?;
    ///     
    ///     // Verify the signature
    ///     let valid = client.sign().verify(&key_id, data, &result.signature, false).await?;
    ///     assert!(valid);
    ///     Ok(())
    /// }
    /// ```
    pub async fn verify(
        &self,
        key_id: &Uuid,
        data: &[u8],
        signature: &[u8],
        prehashed: bool,
    ) -> Result<bool> {
        #[derive(Serialize)]
        struct Request {
            data: String,
            signature: String,
            prehashed: bool,
        }

        #[derive(Deserialize)]
        struct Response {
            valid: bool,
        }

        let request = Request {
            data: BASE64.encode(data),
            signature: BASE64.encode(signature),
            prehashed,
        };

        let response: Response = self
            .client
            .post(&format!("/v1/keys/{}/verify", key_id), &request)
            .await?;

        Ok(response.valid)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sign_client_creation() {
        let client = Client::new("test_key");
        let _sign = client.sign();
        // Just verify it compiles and doesn't panic
    }
}

